WIP refactoring

uellue opened the pull request:
Hi @DavidLanders95 , this is so-to-speak a snapshot of what I am working on.

* Have users interact only with individual values, no arrays or matrices facing to the user
* Use JAX to extract matrices
* Try to reduce boilerplate
* Remove pathlength from ray since that is not linear, and define separate function instead. How to handle this for phase calculations is TBD. Possibly one can also replace it with a phase offset to avoid the issues with floating point precision?

The machinery to switch between nice user-friendly definitions and efficient matrix code seems to work quite well already.

Currently WIP Putting together the code necessary to build the simple 4D STEM model. Next steps are scan and detector grids, and then representing a model

uellue committed:
WIP refactoring

* Have users interact only with individual values
* Use JAX to extract matrices
* Try to reduce boilerplate
* Remove pathlength from ray since that is not linear, and define separate function instead.
  How to handle this for phase calculations is TBD. Possibly one can also replace it with a
  phase offset to avoid the issues with floating point precision?

The machinery to switch between nice user-friendly definitions and efficient matrix code
seems to work quite well already.

Currently WIP Putting together the code necessary to build the simple 4D STEM model.
Next steps are scan and detector grids, and then representing a model

DavidLanders95 commented on src/jaxgym/ray.py:
```diff
@@ -10,6 +11,57 @@
+    # Important to also include here for differentiation to work
+    # Part of the matrix to allow constant offsets.
+    # Don't set this as a user, has to be value 1!
+    _one: float = 1.
```
YES! This is a good solution to the problem of the shifter and my awkward OPL derivative to calculate it.

DavidLanders95 commented on src/jaxgym/ray.py:
```diff
@@ -10,6 +11,57 @@
 
 @jdc.pytree_dataclass
 class Ray:
+    x: float
+    y: float
+    z: float
+    dx: float
+    dy: float
```
I don't quite agree with seperating out the variables like this: The reason the ray.matrix variable existed was because it is the direct column vector that multiplies into the 5x5 Jacobian Ray transfer matrices of components. That matrix was the definitive state vector of the linear system [x, y, dx, dy, 1] and had a variable named after it as a result. Yes I then needed convenience methods to access it's variables, but now we need convenience methods to create this matrix. I suppose it is a pick your poison approach, maybe you can justify it here and we can decide.

DavidLanders95 commented on src/jaxgym/ray.py:
```diff
@@ -10,6 +11,57 @@
+    # Part of the matrix to allow constant offsets.
+    # Don't set this as a user, has to be value 1!
+    _one: float = 1.
+
```
I also don't quite agree with removing path-length from the ray for the following reasons:

1. It is just a parameter of the ray, so it doesn't matter that the values it inherited from components contained at least quadratic terms or higher, which you mention in the pull request as the reason for removing it. I think the same goes for x,y,dx,dy - those are just parameters, and could be made up of non-linear values in the end. 

2. If one knows the path length for a ray, and can calculate it's partial derivatives with respect to input coordinates, one can say what the output coordinates are. The optical path length is the "generating function" for the entire system. If one knows the optical path length of a ray, or the optical path length change that a component applies, and can calculate it's partial derivatives, one can make an equation for the entire system (unless there is diffraction). I think it has a privileged place, and should be a parameter of the ray. See this https://en.wikipedia.org/wiki/Hamiltonian_optics#The_Euler-Lagrange_equations - L here in this equation is the optical path length: From knowing L, and being able to differentiate it, one can obtain the equation to tell you ray coordinates in the system. 

3. Furthermore, one can also calculate the "amplitude" of a ray, by calculating partial derivatives of the optical path length again (and then a determinant operation... ). In one sense, the ray, and components, should only have an optical path length parameter, and nothing else! But that is a bit extreme. 

I suppose, what the "ray" even is is the source of this confusion, as it can represent two things: 

    1. It can be a single perturbation to the system, that allows us to work out the ray transfer matrices (and polynomial coefficients for non linear components). If it was just this, I think it could even only have an optical path length parameter, and with partial derivatives we could obtain the equation (via partial derivatives of the optical path length)
    at any plane. 

    2. It can be a storage device to obtain coordinates from one plane to another. It it was just this, it could have a matrix parameter formed of: (x, y, dx, dy, 1), and a z parameter for convenience. *Although this "state vector" only works for linear components (5x5 ray transfer matrices).

I wrote the old Ray so that it could perform both roles, hence the "matrix" attribute, and the optical path length attribute. I hope this makes sense - it warrants further discussion I think.

uellue replied:
I think we are getting to the core of figuring out the best data structures and API structure. I do understand your points!

The idea here was the following:

* Limit the Ray data structure and step function of components to what we can calculate with a linear function for our basic optical components so that we can ensure equivalency of step function and matrix product implementation, and of Ray and RayMatrix. Everything else should go to different data structures and different functions.
* Minimize the amount of code that translates between values and vectors. In particular, don't let users or component implementations interact with elements of ray vectors or transfer matrices since this makes refactoring a nightmare should we want to change the matrix representation.
* The best data structure for fast ray tracing is a matrix that allows tracing many rays at once. That's why the RayMatrix is defined for arrays of rays and we skip the individual vector representation altogether. Typically this should also make the work of compilers and of differentiation with the Ray data structure easier -- at least with Numba individual numerical values are the winning style for such short vectors.

Looking forward to discussing options to bring the two perspectives together once the PoF schedule allows!

DavidLanders95 replied:
Yes I hope so! 

> * Limit the Ray data structure and step function of components to what we can calculate with a linear function for our basic optical components so that we can ensure equivalency of step function and matrix product implementation, and of Ray and RayMatrix. Everything else should go to different data structures and different functions.

I don't think the rays having a parameter that can be non-linear (quadratic at most in all paraxial cases such as the OPL) gets in the way of ensuring equivalency.  When you form the RayMatrix you use individual parameters anyway so one can just leave out the OPL parameter as part of that. Also as we discussed elsewhere there is a z in there which I do not think belongs - it interferes with the conventional [[A, B], [C, D]] form of the ray transfer matrices.

> * Minimize the amount of code that translates between values and vectors. In particular, don't let users or component implementations interact with elements of ray vectors or transfer matrices since this makes refactoring a nightmare should we want to change the matrix representation.

Ok fair enough, but I don't think we could even do this if we tried! As far as I can understand, to change the matrix representation would mean almost to not use it at all, and to use something else like direction cosines. 

> * The best data structure for fast ray tracing is a matrix that allows tracing many rays at once. That's why the RayMatrix is defined for arrays of rays and we skip the individual vector representation altogether. Typically this should also make the work of compilers and of differentiation with the Ray data structure easier -- at least with Numba individual numerical values are the winning style for such short vectors.

One could still trace many rays at once with the matrix representation I had, one just needed a call to jax's vmap on the first dimension. However, I like that we now have the ray._one variable and can differentiate with respect to individual variables. It means we can't call ray.jacobian any more and expect something that is a 5x5 matrix directly on the ray, but we can still form that matrix at the end if we want to do that. 

I think though for me, definitely the z parameter should be taken out of the ray vector that is formed to calculate the component transfer matrix and I also think opl should be back as it is too useful not to use, and we can just not use it when calculating ray transfer matrices on components. I will play with this tomorrow and add to the pull request my suggestions!

DavidLanders95 commented on tests/test_component.py:
Shouldn't a lens with infinite focal length send all rays out parallel to the optical axis?

DavidLanders95 commented on tests/test_component.py:
```diff
@@ -1,11 +1,308 @@
+
+
+def test_lens_infinity():
+    # A thin lens with infinite focal length
```
Shouldn't a lens with infinite focal length send all rays out parallel to the optical axis?

DavidLanders95 commented on src/jaxgym/components.py:
```diff
@@ -15,104 +17,184 @@
+        sample = Ray(
+            x=1., y=2., z=3., dx=4., dy=5.
+        )
+        # Get the Jacobian which is the transfer matrix
+        mat = jac(unwrap(sample))
```
This jacobian will be 6x6 now? With rows corresponding to those that modify [x, y, z, dx, dy, 1]
I need to think about this some more... this changes the fundamental coordinate system of the ray matrix, with a z now in the middle after x and y. Nothing wrong with it, I was just following conventions I am familiar with, where z was treated almost as a time-like parameter, and not a coordinate of the ray perse.

uellue replied:
The time coordinate should also be related to the path length, right?

If one adds a dz as well, one has a data structure to represent a moving particle in 3D space. I wasn't quite sure what to make of this at this point. Maybe matrix optics as a subset of 3D particle tracking?

DavidLanders95 commented on tests/test_component.py:
Thanks for adding these fundamnetal tests also.

DavidLanders95 commented on src/jaxgym/ray.py:
```diff
@@ -10,6 +11,57 @@
 
 @jdc.pytree_dataclass
 class Ray:
+    x: float
+    y: float
+    z: float
+    dx: float
+    dy: float
+    # Important to also include here for differentiation to work
+    # Part of the matrix to allow constant offsets.
+    # Don't set this as a user, has to be value 1!
+    _one: float = 1.
+
+    def modify(self, x=None, y=None, z=None, dx=None, dy=None) -> "Ray":
+        loc = locals()
+        params = {}
+        for key in ('x', 'y', 'z', 'dx', 'dy'):
+            ll = loc.get(key, None)
+            if ll is None:
+                ll = getattr(self, key)
+            params[key] = ll
+        return Ray(**params, _one=self._one)
```
```suggestion
@jdc.pytree_dataclass
class Ray:
    x: float
    y: float
    dx: float
    dy: float
    # Important to also include here for differentiation to work
    # Part of the matrix to allow constant offsets.
    # Don't set this as a user, has to be value 1!
    _one: float = 1.

```

If we are going to go down the route of individual variables existing only on the Ray object, and always using all of them to wrap and unwrap the matrix representation, then I think this means that z and opl in the most basic form, should not live on the ray object. The fact that z lived there was always only a convenience, and was slightly cursed of me to write it like that. This is because the optical system is parametrised as a function of z, one of 3 possible choices (s - pathlength (NOT optical path length), 
z - distance along the optical axis, and t - time):

Where to find r at the end, we sum over a bunch of step functions to get us there, all parametrised in some form whereby we know their coordinates. 

So then any ray coordinate is a function of one of these parameters:
### 1. Parametrization by \( ds \)
\[
r(s_f) = r(s_i) + \sum_{j=i}^{N} \delta r(s_j)\,\Delta s,
\]

### 2. Parametrization by \( dz \)
\[
r(z_f) = r(z_i) + \sum_{k=i}^{M} \delta r(z_k)\,\Delta z,
\]

### 3. Parametrization by \( dt \)
\[
r(t_f) = r(t_i) + \sum_{l=i}^{L} \delta r(t_l)\,\Delta t,
\]

In all cases, sense we parametrise the model as a function of z we should and will always know the z-positions, and we use those to query the ray object. As a result, z can't live on the ray object, and if it does so it is only as a convenience for us. But then, if it lives in the ray object, it should NOT be part of the vector that we call jacobian on, as this ruins the natural 5x5 matrix that would pop out of a call to the jacobian. Hence the original existence of the matrix variable on the ray. 

I am not sure how to include z and opl seperately, and have the wrap and unwrap functions not need to call more variables so they become verbose.

DavidLanders95 commented on src/jaxgym/components.py:
```diff
@@ -15,104 +17,184 @@
-    @property
-    def z(self):
-        return self.z_po
+@jdc.pytree_dataclass
+class FreeSpace(LinearComponent):
+    length: float
+
+    def step(self, ray: Ray) -> Ray:
+        return ray.modify(
+            x=ray.x + self.length * ray.dx,
+            y=ray.y + self.length * ray.dy,
+            # The multiplication with "one"
+            # is important for differentiation to work
+            z=ray.z + self.length * ray._one,
+        )
+
+    def pathlength(self, ray: Ray) -> float:
+        return self.length * jnp.linalg.norm(
+            jnp.array((1, ray.dx, ray.dy))
+        )
```
In line with my other comments about the parametrisation of the system, when you make a model, you naturally choose components and give them a z, thus defining where free space exists in the model. I don't think we need a specific component for it. Gradients with respect to ray._one or ray.z here are also needless, as we have that information already when we parametrise the system. I think if we want gradients for z positions then we should differentiate with respect to the model component.z.

uellue replied:
> In line with my other comments about the parametrisation of the system, when you make a model, you naturally choose components and give them a z, thus defining where free space exists in the model. I don't think we need a specific component for it. Gradients with respect to ray._one or ray.z here are also needless, as we have that information already when we parametrise the system. I think if we want gradients for z positions then we should differentiate with respect to the model components and their zs, rather than a FreeSpace function

A separate `propagate` function is just another `step` function in disguise IMO. By introducing the FreeSpace component we reduce the size of the API as well as testing efforts.

Filling gaps between components with free space would be a natural job for the model IMO -- not more difficult than deciding when and how to call a propagate function.

Also, in a real microscope the free space is not really free, but should/could be modeled just like any other part of the column as a space filled with electromagnetic fields.

Regarding z, see my other comment! It should be no problem to find another solution.

DavidLanders95 replied:
Ok, fair enough. Nothing more to say here, except that if the free space is not really free, than it should be a component that is modelling it 😃.

DavidLanders95 reviewed (commented):
I think almost all my comments boil down to; There must exist a fundamental vector for the ray somewhere that is [x, y, dx, dy, 1], and even if it is not this, there is no place for a z variable inside it, unless we reparametrise the model as function of a different variable, but then we don't have ray transfer matrices anymore so easily.

DavidLanders95 committed:
Suggestion for simpler ray object that has sensible state vector with no z parameter.

DavidLanders95 reviewed (commented):
Thinking about this a little bit more - if we want the flexibility to have the ray take on whatever parameters we want, and still be able to construct sensible jacobians using [x, y, dx, dy, 1], we can just write a custom partial derivative function to make the 5x5 partial derivative matrix ourself, instead of being restricted to calling jax.jacobian on a ray vector.

DavidLanders95 commented:
> > I think almost all my comments boil down to; There must exist a fundamental vector for the ray somewhere that is [x, y, dx, dy, 1], and even if it is not this, there is no place a z variable inside it, unless we reparametrise the model as function of a different variable, but then we don't have ray transfer matrices anymore so easily.
> 
> Probably we should have a look at how the API would be used downstream, for example for plotting the rays in a microscope model and in the microscope calibration. For the former we need x, y, z of the start and end of the segment. For the latter we actually don't need it, but we are interested in x and y coordinates at two specific z levels -- specimen and detector.
> 
> I'm sure there are many ways to deal with the z coordinate that will work in practice. My goal was to keep the API small, and it looked like z could be tracked the same way as x and y, but that is surely not the only way. Thinking about it, probably the model is also a good place to deal with z -- in the real world it is the microscope column or optical table that puts components in their place. 😊 Conceptually I just find it strange that a ray is not fixed in the direction of the optical axis. However, it can be seen as the model's business to determine where it starts and track it through the various components.

In a way, we are not limited to JAX also dealing with the plotting of rays at all. Jax can give us gradients, and if thats all it does then that can be fine. The model and it's z coordinates, along with the ray starting positions can then be handed off to old temgym or normal python for plotting perhaps. I think trying to make the ray object be a one size fits all object would be nice, but if it gets in the way of clean gradients, then it's not worth it for me.

DavidLanders95 commented:
> Conceptually I just find it strange that a ray is not fixed in the direction of the optical axis. However, it can be seen as the model's business to determine where it starts and track it through the various components.

For certain components the optical axis will do a horse shoe shape (a filter I think), and then z as a parameter might not be single valued, so you need to re-parameterise it then as something else.

DavidLanders95 commented:
Just reflecting on it some more this morning, the problem comes from wanting a specific structure to satisfy convention in optics - i.e the ray vector we call the Jacobian on has to be [x, y, dx, dy, 1]. What if we never called the Jacobian, and abandoned the idea of forming matrices of components, and simply let them be represented by linear functions. The linear matrix is just 5 linear equations anyway, and we can still combine the coefficients of 5 linear equations to transport rays through multiple components. This way, the ray vector can be more flexible, and we simply take partial derivatives of specific variables and form the set of equations this way. It's a convenience to be able to call the Jacobian on a vector [x, y, dx, dy, 1], but that's all. Plus, for higher order derivatives the matrix representation get's messy anyway, and I will be working with a set of polynomial coefficients and will certainly abandon the matrix method. This way z (or any other variable for that matter) can live on the ray happily, and we just ignore derivatives of it when we don't want them.

DavidLanders95 committed:
Added notebook for options

DavidLanders95 commented:
Reflecting on what we discussed yesterday, and thinking about it some more with Matthew, I've added a notebook with 3 options I can think of for the ray dataclass. 

I think if we loosen the criteria that the ray must have an equivalent matrix representation that can be transported back and forth, it gives us more freedom to add whatever parameters we like to the ray. My vote is for something that looks like option 1 in the notebook with a ray like this:

@jdc.pytree_dataclass
class Ray:
    x: float
    y: float
    dx: float
    dy: float
    z: float
    pathlength: float
    _one: float = 1.0

which still has all parameters we could want (or more), can be vectorised, and simply requires us to write a custom jacobian function to make a 5x5 matrix that is sensible in optics. I think that's a small overhead. It means now that a ray will have to form a smaller sub-matrix to do matrix multiplication with the 5x5 matrix of a component, but that is not too much of a problem I don't think.

uellue commented:
> a light optics simulator built in jax

Oh nice! Looks like it is a nice sibling for JaxTemGym. It would be cool to translate between TemGym and Chromatix models automatically.

uellue commented:
Hmm, maybe it is a good time to think about the purpose of RayMatrix and MatrixComponent. My motivation to include them was the following:

* Test code to ensure that differentiating the `step()` function gives a correct transfer matrix so that we can use it to construct more complex functions of the model.
* Possibly propagating rays at scale, although that aspect might be a case of YAGNI and probably actually done by functions extracted from the more complex functions of the model, and not at the Component level. The RayMatrix might be a good input data type for the model, though.

Thinking about it from this perspective, one can argue for "divide and conquer" to make the `Ray` class and `step()` function as simple as possible -- no Z and no optical path length. Instead, these should live in separate functions and data structures that can be tested separately.

From our discussion it could be a good check to ensure the `step()` function is indeed the derivative of the `pathlength()` function.

I'll be back on the 28th of April. I hope today I can still build an example of what a Model could look like!

uellue committed:
WIP refactor the transforms

The overall idea is to express scan and detector coordinates
as such transforms, and allow to derive forward and backward transform
functions between physical coordinates and pixel coordinates
very simply and concisely for later use in the model.

In a later stage one can also use similar approaches for 3D coordinates
to express kinematics of complex actuators, such as specimen stages etc.
That was an issue with interpreting NeXus files from beamlines with such
actuators. At the time we were missing a
handy library to interpret the actuator positions in a meaningful way.
It could become very simple with a composition of transforms that
are validated against a model of the beamline, for exmple in TEMGYM.

Work to make them explicit in x and y coordinate for the user, similar to
the rays and components.

Base them on a simple definition of origin and base vectors since that might
be intuitive for users and also allows for sheared coordinate systems.

Not sure if this is a good approach, though... Just playing around and
seeing how it feels!

The missing transform functions can be added bit by bit from the original file.

uellue commented:
Got a bit stuck with the coordinate transforms. :-)

uellue committed:
Save previous work

Migrating to new version by David

DavidLanders95 commented:
Hi Dieter, just to summarise our discussion earlier, I thought I'd make a comment here and add a few points after Matthew and I talked about it a bit more, which are worth mentioning also. On the first point - 

1. Components in the pytree_dataclasses should just become functions

Right now yes in the microscope-calibration model they don't offer much and could be functions, however I forgot to mention when we came to try to bring back the GUI from temgym last week, and find a way to populate a GUI with parameters made from a JaxGym model, it was very easy to do so because the components were dataclasses.
We could just loop through the jax_dataclasses and get the names and values of parameters
```
for component in components:
for name, val in asdict(component).items():
])
```
giving us this: 
<img width="2547" height="1129" alt="image" src="https://github.com/user-attachments/assets/2a2d64d3-f0e4-45cd-adee-0d62c18ddc15" />

Not imperative, and there are lots of ways to do it, but it is convenient for some extra functionality that we sometimes like to have. At some point one is going to need to make a NamedTuple of parameters somewhere anyway so I personally still think jax dataclasses are an ok solution, and doesn't add much extra code for something that is by default immutable, which jax likes, and gives you some extra functionality in case you need it. One compromise is to have something like this perhaps but it still has the boiler plate of the dataclass.

```
def lens(ray: Ray, z: float, focal_length: float) -> Ray:

    f = focal_length

    x, y, dx, dy = ray.x, ray.y, ray.dx, ray.dy

    new_dx = -x / f + dx
    new_dy = -y / f + dy

    pathlength = ray.pathlength - (x**2 + y**2) / (2 * f)
    one = ray._one * 1.0

    return Ray(
            x=x, y=y, dx=new_dx, dy=new_dy, _one=one, pathlength=pathlength, z=ray.z
        )


@jdc.pytree_dataclass
class Lens:
    z: float
    focal_length: float

    def step(self, ray: Ray):
        return lens(ray, *self)
```
So maybe there is another solution, but using Jax dataclasses by default implies components are immutable, which is a bit more explicit and plays nicely with Jax. If we use raw NamedTuples, then I don't think we get the ability to subclass like we do with the GridBase class for coordinate transforms. If we make components functions, then the NamedTuple of inputs should live somewhere anyway! 

2. It should be easier to choose what parameters to differentiate with respect to. 
Fully agree, and although there are lots of solutions to this, the simplest is still that one should write a wrapper to choose what parameters one wants to differentiate with respect to. In a way, we don't explicitly show this being done in microscope calibration, but it is possible do do it for coordinate transforms too, however yes it's not made very easy, which leads me to the third point:

3. Coordinate Transforms to pixels should be more easily differentiable:

As I said at the moment this is possible still via a wrapper function, and one does have the box of lego to achieve this, however it's not very convenient and it could be made moreso. 
I think one place that does show the advantages of your idea is here:
```
@jax.jit
def project_coordinates_backward(
    model: Model, det_coords: np.ndarray, scan_pos: Coords_XY
) -> np.ndarray:
    PointSource = model.source
    ScanGrid = model.scan_grid
    Detector = model.detector
    semi_conv = PointSource.semi_conv

    # Return all the transfer matrices necessary for us to propagate rays through the system
    # We do this by propagating a single ray through the system, and finding it's gradients
    _, total_transfer_matrix, det_to_scan = solve_model_fourdstem_wrapper(
        model, scan_pos
    )

    # Get ray coordinates at the scan from the det
    scan_rays_x, scan_rays_y, detector_mask = ray_coords_at_plane(
        semi_conv,
        scan_pos,
        det_coords,
        total_transfer_matrix,
        det_to_scan,
        Detector.det_pixel_size,
    )

    # Convert the ray coordinates to pixel indices.
    scan_y_px, scan_x_px = ScanGrid.metres_to_pixels([scan_rays_x, scan_rays_y])

    return scan_y_px, scan_x_px, detector_mask
```
Matthew and I thought about and wrote some pseudocode, and if we added some extra components that did coordinate transforms, or found a way for it to be done on the ScanGrid and Detector somehow:

```
@jax.jit
def project_coordinates_backward(
    model: Model, det_coords: np.ndarray, scan_pos: Coords_XY
) -> np.ndarray:
    PointSource = model.source
    ScanGridMetresToPixels = model.coordinate_transform
    ScanGrid = model.scan_grid
    DetectorMetresToPixels = model.coordinate_transform
    Detector = model.detector
```
one could then conveniently choose one of those planes in the model to get ray coordinates in pixels, could definitely more easily write a wrapper to differentiate with respect to parameters in those spaces, and it would also mean that potentially with extra work (and somehow a blocked parameter, TBD) we could remove or simplify these lines altogether. 

```
    # Get ray coordinates at the scan from the det
    scan_rays_x, scan_rays_y, detector_mask = ray_coords_at_plane(
        semi_conv,
        scan_pos,
        det_coords,
        total_transfer_matrix,
        det_to_scan,
        Detector.det_pixel_size,
    )

    # Convert the ray coordinates to pixel indices.
    scan_y_px, scan_x_px = ScanGrid.metres_to_pixels([scan_rays_x, scan_rays_y])
```

So that would be a nice change for sure, and make the whole model more flexible to differentiate with respect to parameters in a different coordinate system, like one might like to do for parameters on the Sample and Detector plane.

After writing all this out, each issue seems to be related in one form or another to coordinate transforms, and how they have been separated from the model transformations. Maybe that helps to zone in on which is the biggest problem to solve first.

matbryan52 commented:
I would just add to @DavidLanders95 's comments that, there are a few clear features we can add that will add to the toolbox of `JaxGym`, to make it easier to do more things within specific applications like `MicroscopeCalibration`.

1. We can go slightly further with the idea of making a component into just a function, something like:

```python
from jaxgym import component

def lens(ray: Ray, z,: float, f: float):
    # The lens as a pure function
    ...

@component(lens)
class Lens:
    z: float
    f: lens
```

where the `@component(fn)` decorator would create the current `step` function, and allow us to hold a model, it's functionality and its parameters as an immutable object. This brings advantages for building the UI, and for Jax in that we can rely on `jdc` or some other similar backend to mark certain parameters differentiable or static etc.

2. Differentiability of any parameter in the model is something we are definitely missing, mainly because Jax support for differentiating in a nested argument is poor. We just need the machinery to point at parameters in the tree that we want to differentiate with respect to. A few libraries already do this, we just need to take one and adapt it to our parameter tree. Something like:

```python
model = (
    comp.Source(z=0.),
    comp.Lens(z=0.5, f=1.),
    comp.Detector(z=1., shape=(256, 256), px_size=0.001),
)
diff_with = (model[1].params.f, model[2].params.z)
ouptut_ray, (grads_f, grads_z) = run_with_grads(input_ray, model, diff_with)   # can also return grads wrt input ray attributes
```

where `params` would be created by `@component` to let us point to specific parameters to differentiate wrt.

3. Coordinate transforms in the model - @DavidLanders95 and I discussed this, and there are a number of ways to make it happen. We can easily add a way for the rays that _go-in-to_ or _come-out-of_ a model are in a different coordinate system (so that they can be differentiated), probably with some kind of `apply_transform` parameter on the gridded components, but if there are coordinate transforms **within** a model tree, at intermediate locations, then we have to be very careful to maintain the main coordinate system so that the rest remains valid. The only case I can see for a transform within a model is if we have a function which can output rays from multiple planes at once.
